File: test/level0/ticket.test.js
// The ticket door. Each case hands the door one write and reads the finding it
// answers, because a case asserting nothing passes whatever the door does.
// [[spec/design_output/schema#the-three-places]]

import assert from "node:assert/strict";
import { test } from "node:test";
import { readNote, readYaml } from "../../.claude/skills/level0/lib/schema.js";
import {
  engineRows,
  placesIn,
  refusedTicket,
  queueHolds,
  ticketFaults,
} from "../../.claude/skills/level0/lib/ticket.js";

const SCHEMA_TEXT = `
kind: ticket

frontmatter:
  type: object
  required:
    - kind
  properties:
    kind:
      const: ticket
      x-link: true
      description: the schema this note is minted from

    state:
      enum: [draft, open, closed]
      x-engine: true
      description: whether anybody pulls it

    step:
      type: string
      x-engine: true
      description: the leaf this ticket stands on

    steps:
      type: array
      x-engine: true
      description: the route

    urgency:
      enum: [now, soon, whenever]
      description: which ticket the pull hands out first

body:
  headingLevel: 1

  sections:
    - header: Ask
      required: true
      x-written: anyone
      description: what this ticket asks for

    - x-one-per: steps
      x-written: hand

    - header: Discussion
      required: true
      position: last
      x-written: anyone
      description: what anybody adds
`;
const SCHEMA = readYaml(SCHEMA_TEXT);

const WHERE = "spec/tickets/a-name.md";

const open = `---
kind: [[ticket]]
state: open
urgency: now
step: implement/change
steps:
  - name: design
    does: writes the design
  - name: implement
    steps:
      - name: change
        does: makes the change
        evidence:
          - name: lint
            form: command
            says: the tree builds and lints
          - name: seen
            form: text
            says: what surprises you
---

# Ask

What it asks for.

# design

# implement

## change

### lint

### seen

# Discussion

Nothing yet.
`;

const weighed = (now, was = open) => ticketFaults(was, now, SCHEMA, WHERE);
const rules = (now, was) => weighed(now, was).map((one) => one.rule);

// [[spec/design_output/schema#the-three-places]]
test("a write to the field of the step in hand stands", () => {
  assert.deepEqual(weighed(open.replace("### lint\n", "### lint\n\nnpm test\n")), []);
  assert.deepEqual(
    weighed(open.replace("### seen\n", "### seen\n\nThe fix is small.\n")),
    [],
  );
});

// [[spec/design_output/schema#the-three-places]]
test("a write to the discussion stands, at any state", () => {
  assert.deepEqual(weighed(open.replace("Nothing yet.", "One thing stands open.")), []);
});

// [[spec/design_output/schema#the-three-places]]
test("a write to the ask stands at any state, so a hand fixes what the rules refuse", () => {
  assert.deepEqual(weighed(open.replace("What it asks for.", "What it asks for now.")), []);

  const draft = open.replace("state: open", "state: draft");
  assert.deepEqual(
    ticketFaults(
      draft,
      draft.replace("What it asks for.", "What it asks for now."),
      SCHEMA,
      WHERE,
    ),
    [],
  );
});

test("a section written by one state alone refuses the others", () => {
  const drafted = readYaml(String(SCHEMA_TEXT).replace("x-written: anyone\n      description: what this ticket asks for", "x-written: draft\n      description: what this ticket asks for"));
  const said = open.replace("What it asks for.", "What it really asks for.");
  const found = ticketFaults(open, said, drafted, WHERE);
  assert.deepEqual(found.map((one) => one.rule), ["Ticket.Ask"]);
  assert.match(found[0].message, /is the engine's to write/);
});

// [[spec/design_output/schema#the-three-places]]
test("a write to a phase's chapter is refused", () => {
  const said = open.replace("# design\n", "# design\n\nA line nobody asks for.\n");
  assert.deepEqual(rules(said), ["Ticket.design"]);
});

// [[spec/design_output/schema#the-three-places]]
test("two leaves naming one field alike stand apart, so a write beside them passes", () => {
  const twice = open
    .replace("# implement\n", "# implement\n\n## first\n\n### tests\n\n<!-- the first -->\n")
    .replace("# Discussion", "## second\n\n### tests\n\n<!-- the second -->\n\n# Discussion");
  assert.deepEqual(weighed(twice, twice), []);
  assert.deepEqual(weighed(twice.replace("### lint\n", "### lint\n\nnpm test\n"), twice), []);
});

// [[spec/design_output/schema#the-three-places]]
test("a write to the field of another leaf is refused", () => {
  const other = open.replace("step: implement/change", "step: design");
  const said = other.replace("### lint\n", "### lint\n\nnpm test\n");
  assert.deepEqual(rules(said, other), ["Ticket.lint"]);
});

// [[spec/design_output/schema#the-verbs-own-their-fields]]
test("an edit to a field the verbs own is refused, and the line points at it", () => {
  for (const [was, now, rule, line] of [
    ["state: open", "state: closed", "Ticket.state", 3],
    ["step: implement/change", "step: design", "Ticket.step", 5],
    ["    does: makes the change", "    does: makes it", "Ticket.steps", 6],
  ]) {
    const found = weighed(open.replace(was, now));
    assert.deepEqual(
      found.map((one) => one.rule),
      [rule],
      `${now} reaches ${rule}`,
    );
    assert.equal(found[0].line, line);
  }
});

// [[spec/design_output/schema#the-verbs-own-their-fields]]
test("a field no rule marks stands, so a person's fields reach no door", () => {
  assert.deepEqual(weighed(open.replace("urgency: now", "urgency: soon")), []);
});

// [[spec/design_output/schema#the-three-places]]
test("a file nothing stands in yet is a mint, and the door asks it nothing", () => {
  assert.deepEqual(ticketFaults("", open, SCHEMA, WHERE), []);
});

// [[spec/design_output/schema#the-three-places]]
test("the places answer the ask, the step's fields and the discussion, at any state", () => {
  assert.deepEqual(
    [...placesIn(readNote(open), SCHEMA).keys()],
    ["1 Ask", "3 lint", "3 seen", "1 Discussion"],
  );
  assert.deepEqual(
    [...placesIn(readNote(open.replace("state: open", "state: draft")), SCHEMA).keys()],
    ["1 Ask", "3 lint", "3 seen", "1 Discussion"],
  );
});

// [[spec/design_output/schema#the-three-places]]
test("a ticket standing at no leaf of its route offers a hand no field", () => {
  const said = open.replace("step: implement/change", "step: nowhere");
  assert.deepEqual([...placesIn(readNote(said), SCHEMA).keys()], ["1 Ask", "1 Discussion"]);
});

// [[spec/design_output/schema#the-three-places]]
test("a ticket with no step stands at its first leaf, so its fields open to the hand the pull gives it to", () => {
  const said = open
    .replace("step: implement/change\n", "")
    .replace("  - name: design\n    does: writes the design\n", "");
  assert.deepEqual(
    [...placesIn(readNote(said), SCHEMA).keys()],
    ["1 Ask", "3 lint", "3 seen", "1 Discussion"],
    "the first leaf's fields stand open",
  );
});

const recorded = open.replace(
  "steps:\n",
  `record:
  - step: implement/change
    hand: a box, a session and an agent
    hash_before: abc1234
    hash_after: def5678
    returns: 1
    answered:
      - name: lint
        exit: 0
        said: the tree lints
steps:
`,
);

// [[spec/design_output/schema#the-record-draws-itself]]
test("the record draws a line under its leaf, and one under a command field", () => {
  assert.deepEqual(engineRows(readNote(recorded), SCHEMA), [
    {
      key: "2 change",
      said:
        "The hand is `a box, a session and an agent`, and the branch runs `abc1234` to `def5678`, and this leaf returns 1.",
    },
    {
      key: "3 lint",
      said: "answered: exit `0`, and the last line reads `the tree lints`.",
    },
  ]);
});

// [[spec/design_output/schema#the-record-draws-itself]]
test("a skipped leaf draws the reason the pull passes it over", () => {
  const said = recorded
    .replace("    returns: 1", "    skipped: true\n    why: the box runs on a desk")
    .replace(/ {4}answered:\n(?: {6}.*\n| {8}.*\n)+/, "");
  assert.deepEqual(engineRows(readNote(said), SCHEMA), [
    { key: "2 change", said: "The pull skips this leaf, because the box runs on a desk." },
  ]);
});

// [[spec/design_output/schema#the-record-draws-itself]]
test("an answered line the record says stands, and one it fails to say refuses", () => {
  const drawn = recorded.replace(
    "### lint\n",
    "### lint\n\nanswered: exit `0`, and the last line reads `the tree lints`.\n",
  );
  assert.deepEqual(ticketFaults(recorded, drawn, SCHEMA, WHERE), []);

  const claimed = recorded.replace("### lint\n", "### lint\n\nanswered: exit `0`.\n");
  const found = ticketFaults(recorded, claimed, SCHEMA, WHERE);
  assert.deepEqual(
    found.map((one) => one.rule),
    ["Ticket.lint"],
  );
  assert.match(found[0].message, /An answered line is the engine's/);
});

// [[spec/design_output/schema#the-record-draws-itself]]
test("an answered line on a leaf with no record entry refuses", () => {
  const claimed = open.replace("### lint\n", "### lint\n\nanswered: exit `0`.\n");
  assert.deepEqual(
    ticketFaults(open, claimed, SCHEMA, WHERE).map((one) => one.rule),
    ["Ticket.lint"],
  );
});

// [[spec/design_output/schema#the-three-places]]
test("the refusal names the finding and the three places", () => {
  const said = refusedTicket(
    WHERE,
    "ticket",
    weighed(open.replace("state: open", "state: closed")),
  );
  assert.match(
    said,
    /^The ticket door refuses this write to spec\/tickets\/a-name\.md\./,
  );
  assert.match(said, /Ticket\.state/);
  assert.match(
    said,
    /the ask, the fields of the step it holds, and the discussion/,
  );
});

// [[spec/design_output/stop#the-mechanical-checks]]
test("the queue holds work where a free open ticket has a leaf a hand takes, or a group reads now", () => {
  const free = (state, by = "anyone", more = "") =>
    `---\nkind: [[ticket]]\nstate: ${state}\nurgency: soon\n${more}steps:\n  - name: do\n    by: ${by}\n---\n\n# Ask\n\nA thing.\n`;
  assert.equal(queueHolds([free("open")]), true, "an open free ticket");
  assert.equal(queueHolds([free("draft")]), false, "a draft waits for its open");
  assert.equal(queueHolds([free("closed")]), false, "a closed one is done");
  assert.equal(queueHolds([free("open", "person")]), false, "a person step is nobody's on this box");
  assert.equal(queueHolds([free("open", "anyone", "group: some-group\n")]), false, "a child rides its group");
  const group = (urgency) =>
    `---\nkind: [[ticket]]\nstate: open\nurgency: ${urgency}\nprocess: [[spec/processes/group]]\nsteps:\n  - name: split\n---\n\n# Ask\n\nA group.\n`;
  assert.equal(queueHolds([group("soon")]), false, "a group at soon is the cloud's");
  assert.equal(queueHolds([group("now")]), true, "a group at now is the desk's");
  assert.equal(queueHolds([]), false);
});

File: src/index/ticket_test.go
// The tickets, over a tree a case writes. A group's standing comes off its
// record, a child reads its group's, and a loose ticket carries none.
// [[spec/guidance/code/testing]]
package main

import (
	"testing"
)

const heldGroup = `---
kind: [[ticket]]
state: open
urgent: true
process: [[spec/processes/group]]
step: children
steps:
  - name: sync
    does: takes trunk into the branch
  - name: children
    by: children
record:
  - step: sync
    hand: box one
    hash_before: aaa
    hash_after: bbb
  - step: children
    hand: box one
    hash_before: bbb
---

# Ask

<!-- goal, as text: what these tickets add up to -->
Two tickets that land as one.

# children

# Discussion
`

const doneGroup = `---
kind: [[ticket]]
state: closed
process: [[group]]
record:
  - step: children
    hash_before: ccc
    hash_after: ddd
---

# Ask

A group already landed.

# Discussion
`

func child(group, state string) string {
	return "---\nkind: [[ticket]]\nstate: " + state + "\ngroup: " + group +
		"\nprocess: [[spec/processes/trivial]]\nstep: do\ntodo: true\nsteps:\n  - name: do\n---\n\n# Ask\n\nOne piece of it.\n\n# do\n\n# Discussion\n"
}

func ticketTree(t *testing.T) string {
	t.Helper()
	root := tree(t)
	write(t, root, "spec/tickets/one-group.md", heldGroup)
	write(t, root, "spec/tickets/a-child.md", child("one-group", "open"))
	write(t, root, "spec/tickets/old-group.md", doneGroup)
	write(t, root, "spec/tickets/an-old-child.md", child("old-group", "closed"))
	write(t, root, "spec/tickets/a-loose-one.md", "---\nkind: [[ticket]]\nstate: open\nsteps:\n  - name: do\n---\n\n# Ask\n\nA ticket in no group.\n\n# Discussion\n")
	return root
}

func byName(said []Ticket) map[string]Ticket {
	out := map[string]Ticket{}
	for _, one := range said {
		out[one.Name] = one
	}
	return out
}

// [[spec/design_output/index#the-index-answers-the-tickets]]
func TestTheIndexAnswersEveryTicketWithItsFields(t *testing.T) {
	db := opened(t, ticketTree(t))
	said, err := Tickets(db)
	if err != nil {
		t.Fatal(err)
	}
	if len(said) != 5 {
		t.Fatalf("five tickets stand, and the index answers %d", len(said))
	}
	rows := byName(said)
	group := rows["one-group"]
	if group.Route != "group" || group.Step != "children" || !group.Urgent || group.State != "open" {
		t.Fatalf("the group answers its fields, and reads %+v", group)
	}
	if group.Says != "Two tickets that land as one." {
		t.Fatalf("the ask's first line stands past the mint's comment, and reads %q", group.Says)
	}
	held := rows["a-child"]
	if held.Group != "one-group" || held.Route != "trivial" || !held.Todo || held.Path != "spec/tickets/a-child.md" {
		t.Fatalf("a child answers its fields, and reads %+v", held)
	}
	if rows["a-loose-one"].Step != "" || rows["a-loose-one"].Route != "" {
		t.Fatalf("a ticket naming no step and no route answers neither, and reads %+v", rows["a-loose-one"])
	}
}

// [[spec/design_output/index#the-index-answers-the-tickets]]
func TestATicketCarriesTheTimeItsFileLastChanged(t *testing.T) {
	db := opened(t, ticketTree(t))
	said, err := Tickets(db)
	if err != nil {
		t.Fatal(err)
	}
	held := byName(said)["a-child"]
	if held.Changed == 0 {
		t.Fatal("a ticket carries the time its file last changed, and this one carries none")
	}
	var stored int64
	if err := db.QueryRow(`SELECT mtime FROM file WHERE path = ?`, held.Path).Scan(&stored); err != nil {
		t.Fatal(err)
	}
	if held.Changed != stored {
		t.Fatalf("the ticket carries the mtime the file table stores, %d, and reads %d", stored, held.Changed)
	}
}

// [[spec/design_output/work#held-derives-from-the-record]]
func TestAStandingReadsOffTheGroupsRecordThroughTheTicket(t *testing.T) {
	db := opened(t, ticketTree(t))
	said, err := Tickets(db)
	if err != nil {
		t.Fatal(err)
	}
	rows := byName(said)
	if rows["one-group"].Standing != standingHeld {
		t.Fatalf("a record entry carrying hash_before and no hash_after holds the group, and it stands at %q", rows["one-group"].Standing)
	}
	if rows["a-child"].Standing != standingHeld {
		t.Fatalf("a child stands where its group's branch stands, and it stands at %q", rows["a-child"].Standing)
	}
	if rows["old-group"].Standing != standingDone || rows["an-old-child"].Standing != standingDone {
		t.Fatal("a closed group stands done, and its child with it")
	}
	if rows["a-loose-one"].Standing != "" {
		t.Fatalf("a ticket in no group carries no standing, and reads %q", rows["a-loose-one"].Standing)
	}
}

// [[spec/design_output/work#held-derives-from-the-record]]
// A todo names the row the ticket stands before, or reads true, and either lights the flag. [[spec/design_output/pull#the-queue-is-an-outline]]
func TestATodoNamingARowLightsTheFlag(t *testing.T) {
	t.Parallel()
	for said, want := range map[string]bool{"true": true, "a-loose-one": true, `"a-loose-one"`: true, "false": false, "": false} {
		if todoIn(said) != want {
			t.Fatalf("todo %q reads %v", said, want)
		}
	}
}

func TestAGroupNobodyHoldsStandsAtTodo(t *testing.T) {
	if heldIn("state: open\nrecord:\n  - step: sync\n    hash_before: aaa\n    hash_after: bbb\n") {
		t.Fatal("an entry with both hashes holds nothing")
	}
	if heldIn("state: open\nsteps:\n  - name: do\n") {
		t.Fatal("a ticket with no record holds nothing")
	}
	if !heldIn("record:\n  - step: sync\n    hash_before: aaa\nurgent: true\n") {
		t.Fatal("an open entry holds the group, whatever key follows the record")
	}
	if groupStanding("open", "steps:\n  - name: do\n") != standingTodo {
		t.Fatal("a group nobody holds stands at todo")
	}
}
